src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableRequestSessionChannel.cs

```diff
@@ -0,0 +1,842 @@
+        }
+
+        private class AsyncRequest : IReliableRequest, IAsyncRequest
+        {
+            private bool _aborted = false;
+            private bool _completed = false;
+            private TaskCompletionSource<object> _tcs;
+            private bool _faulted = false;
+            private TimeSpan _originalTimeout;
+            private Message _reply;
+            private ReliableRequestSessionChannel _parent;
+
+            public AsyncRequest(ReliableRequestSessionChannel parent)
+            {
+                _parent = parent;
+            }
+
+            private object ThisLock { get; } = new object();
+
+            public void Abort(RequestChannel channel)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _aborted = true;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                    }
+                }
+            }
+
+            public void Fault(RequestChannel channel)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _faulted = true;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                    }
+                }
+            }
+
+            public void Complete()
+            {
+            }
+
+            public async Task SendRequestAsync(Message message, TimeoutHelper timeoutHelper)
+            {
+                _originalTimeout = timeoutHelper.OriginalTimeout;
+                if (!await _parent.connection.AddMessageAsync(message, timeoutHelper.RemainingTime(), this))
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(_parent.GetInvalidAddException());
+            }
+
+            public void Set(Message reply)
+            {
+                lock (ThisLock)
+                {
+                    if (!_completed)
+                    {
+                        _reply = reply;
+                        _completed = true;
+                        _tcs?.TrySetResult(null);
+                        return;
+                    }
+                }
+                if (reply != null)
+                {
+                    reply.Close();
+                }
+            }
+
+            public async Task<Message> ReceiveReplyAsync(TimeoutHelper timeoutHelper)
+            {
+                bool throwing = true;
+
+                try
+                {
+                    bool expired = false;
+
+                    if (!_completed)
+                    {
+                        bool wait = false;
+
+                        lock (ThisLock)
+                        {
+                            if (!_completed)
+                            {
+                                wait = true;
+                                _tcs = new TaskCompletionSource<object>();
```

Comment:
Use TaskCreationOptions.RunContinuationsAsynchronously
Consider using non-generic TaskCompletionSource

Reply:
Can't use non-generic TCS due to targeting netstandard2.0. I'll fix the constructor.